docs/source/models/spatial-knn.rst

```diff
@@ -0,0 +1,288 @@
+the smallest distance. ST_Distance function is defined as the shortest distance between two geometries in projected units.
+We do not restrict the type of geometries that can be used in the problem. The only requirement is that the geometries
+must be in the same coordinate system.
+
+Spatial KNN as a Join Relation
+##############################
+
+The traditional definition of the nearest neighbour problem is defined as:
+"given a set S of points in a space M and a query point q in M, find the point in S that is closest to q".
+We have relaxed this definition by allowing the query point to be a set of points. The result of the nearest neighbour
+problem is a set of points from S that are closest to each query points. This in effect formulates a join condition
+where a point pair (Q[i], S[j]) is a match if st_distance(Q[i], S[j]) <= st_distance(Q[i], S[k]) where S[k] is the
+kth nearest neighbour of Q[i].
+
+The problem of finding the k-nearest neighbours can be formulated as a join condition where a point pair (Q[i], S[j]) is
+the fact that this is not an equality relation and the join condition is not symmetric. If S[j] is in the knn set of Q[i]
+it does not imply that Q[i] is in the knn set of S[j]. This join would default to a Cartesian product if we were to
+use the traditional join syntax. We have to use a different approach.
+
+Spatial KNN as an Iterative Transformation
+##########################################
+
+The problem of finding the k-nearest neighbours can be formulated as an iterative transformation. The transformation
+is defined as follows:
+* For each geometry in set Q generate a hex ring in grid index space
+* Generate match candidates in the hex ring
+* For each match candidate calculate the distance to the query geometry
+* For each geometry in set Q count the matches and stop if the count is equal to k
+* Continue with the next hex ring until max number of iterations is performed
+* Return the geometries that have the smallest distance to the query geometries
+
+.. figure:: ../images/spatial_knn_iterations.png
+   :figclass: doc-figure
+
+   Fig 2. Spatial KNN example over 4 iterations.
+
+Usage
+#####
+
+Mosaic implements a transformer that implements the iterative approach outlined above.
+The transformer is called SpatialKnn and it is used as follows:
+
+.. tabs::
+   .. code-tab:: py
+
+    >>> import mosaic as mos
+    >>> mos.enable_mosaic(spark, dbutils)
+    >>>
+    >>> buildings_df = spark.read.table("buildings")
+    >>> buildings_df = buildings_df.withColumn("left_id", monotonically_increasing_id())
+    >>> roads_df = spark.read.table("roads")
+    >>> roads_df = roads_df.withColumn("right_id", monotonically_increasing_id())
+    >>>
+    >>> from mosaic.models import SpatialKnn
+    >>> knn = SpatialKNN(roads_df)
+    >>> knn.setLeftGeometryColumn("geometry")
+    >>> knn.setRightGeometryColumn("geometry")
+    >>> knn.setLeftIdColumn("left_id")
```

Comment:
Left/Right are not very clear.
Which one is S and which one is Q? Do you have some other candidate names? Q is usually the POI, right?